File: c++/questions/NQueens.cpp
#include <vector>
#include <string>
#include <unordered_set>

using namespace std;

class Solution {
public:
/**
 * @brief N queens O(n!) solution. O(n^2) space solution. As the tree gets larger (n), the number of options began to decrease hence O(n!)
 * 
 * @param n size of the board
 * @return vector<vector<string>> the possible solutions
 */
    vector<vector<string>> solveNQueens(int n) {
        this->n = n;
        board = vector<vector<char>>(n, vector<char>(n, '.'));
        unordered_set<int> col, leftDiag, rightDiag;
        backTrack(0, col, leftDiag, rightDiag);
        return res;

    }
private:
    int n = 0;
    vector<vector<char>> board;
    vector<vector<string>> res;


    /**
     * @brief Add valid solutions to res
     * 
     * @param r current row
     * @param vert invalid verticals
     * @param leftDiag invalid left diagonals
     * @param rightDiag invalid right diagonals
     */
    void backTrack(int r, unordered_set<int>& vert, 
    unordered_set<int>& leftDiag, 
    unordered_set<int>& rightDiag){

        if (r == n){
            vector<string> copy;
            for (auto & col: board){
                string row(col.begin(), col.end()); // requires col to be vector<char> rather than string
                copy.push_back(row);
            }
            res.push_back(copy);
            return;
        }

        for (int c = 0; c < n; c++){
            if (vert.count(c) == 0 && leftDiag.count(r + c) == 0 && rightDiag.count(r - c) == 0){
                board[r][c] = 'Q';
                vert.insert(c); leftDiag.insert(r + c); rightDiag.insert(r - c);
                
                backTrack(r + 1, vert, leftDiag, rightDiag);

                board[r][c] = '.';
                vert.erase(c); leftDiag.erase(r + c); rightDiag.erase(r - c);
            }
        }
    }   
};

File: Python/Questions/NQueens.py
from typing import List
class Solution:
    def solveNQueens(self, n: int) -> List[List[str]]:
        """ N queens O(n!) solution. O(n^2) space solution. As the tree gets larger (n), the number of options began to decrease hence O(n!)

        Args:
            n (int): The size of the chess board.

        Returns:
            List[List[str]]: Valid possible solutions to the board.
        """
        res, board = [], [["."] * n for i in range(n)]
        def backtrack(r, vert, rightDiag, leftDiag):
            if r == n:
                copy = ["".join(col) for col in board]
                res.append(copy)
                return
            
            for c in range(n):
                if c not in vert and r - c not in rightDiag and r + c not in leftDiag:
                    board[r][c] = "Q"
                    vert.add(c)
                    rightDiag.add(r-c)
                    leftDiag.add(r+c)
                    backtrack(r + 1, vert, rightDiag, leftDiag)

                    board[r][c] = "."
                    vert.remove(c)
                    rightDiag.remove(r-c)
                    leftDiag.remove(r+c)
            
        backtrack(0, set(), set(), set())
        return res

        # res = []
        # def dfs(i, pos, invalid):
        #     if i == n:
        #         res.append(pos.copy())
        #         return 
            
        #     for j in range(n):
        #         if (i, j) not in invalid:
        #             newInvalid = invalid.copy()
        #             # add vals to invalid

        #             # add vertical cols
        #             for r in range(i + 1, n):
        #                 newInvalid.add((r, j))
        #             # add diagonals
        #             for d in range(1, n):
        #                 newInvalid.add((i + d, j + d))
        #                 newInvalid.add((i + d, j - d))

        #             pos.append(f'{"." * j}Q{"." * (n - j - 1)}')
        #             dfs(i + 1, pos, newInvalid)
        #             pos.pop()
        
        # dfs(0, [], set())
        # return res



        